src/api/index.spec.tsx

```diff
@@ -773,9 +845,57 @@ describe(__filename, () => {
       expect(_callApi).toHaveBeenCalledWith({
         apiState,
         endpoint: `reviewers/addon/${addonId}/versions/${versionId}/draft_comments`,
+        endpointUrl: undefined,
         method: HttpMethod.GET,
       });
     });
+
+    it('can get comments with an endpointUrl', async () => {
+      const endpointUrl = 'https://example.com/endpoint/?page=2';
+      const _callApi = jest
+        .fn()
+        .mockResolvedValue(
+          createFakeCommentsResponse([createFakeExternalComment()]),
+        );
+
+      await _getComments({ _callApi, endpointUrl });
+
+      expect(_callApi).toHaveBeenCalledWith(
+        expect.objectContaining({ endpoint: undefined, endpointUrl }),
+      );
+    });
+  });
+
+  describe('getAllComments', () => {
+    it('calls fetchAllPages', async () => {
+      const _getComments = jest.fn();
+      const _fetchAllPages = jest.fn();
+      const addonId = nextUniqueId();
+      const apiState = defaultApiState;
+      const versionId = nextUniqueId();
+
+      await getAllComments({
+        _fetchAllPages,
+        _getComments,
+        addonId,
+        apiState,
+        versionId,
+      });
+
+      expect(_fetchAllPages).toHaveBeenCalled();
+
+      const getNext = _fetchAllPages.mock.calls[0][0];
+
+      const nextUrl = 'endpoint/?page=2';
+      getNext(nextUrl);
```

Comment:
This feels a bit odd to me. IIUC, `getNext` is the function that would have been called by `_fetchAllPages`? So what you're doing is taking that function and calling it with an expected value so that you can then assert that it did what you expected it to do (i.e., call `_getComments` with the expected value). This all seems to make sense, except for the fact that you're manually calling `getNext`. Shouldn't `_fetchAllPages` have already called `getNext`, and if so wouldn't it be nicer if we could just assert about what happened at that time, rather than having to manually call `getNext` ourselves?

I gather this is being done because of the complicated way this all works under the covers which makes it tricky to test, but I thought I'd raise it to see if it's possible to do it the way I'm suggesting above.

Reply:
I added a comment to make this test code more clear.

It's treating `fetchAllPages` as an external dependency with a known interface: a function taking a callback parameter. It is safe to mock this because TS will enforce adherence to the interface.

The test then simulates the interface which makes the test a lot simpler. It doesn't have to set up a mock response that returns a next page URL and doesn't have to worry about avoiding infinite fetching loops. If the implementation of `fetchAllPages` were to later change, this test would not break as long as the interface didn't change.

Reply:
💡 I moved this code to a helper since it's likely that other API tests would need it in the future. This should make the test code a little easier to understand, as well. I added more type safety to it so that the test code will produce an error like this for incorrect usage:

```
src/api/index.spec.tsx:906:7 - error TS2554: Expected 1 arguments, but got 0.

906       getNextResponse();
          ~~~~~~~~~~~~~~~~~

  src/api/index.tsx:264:5
    264     nextUrl: PaginatedResponse<ResultsType[]>['next'],
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'nextUrl' was not provided.
```